File: lobster-input-ios/LobsterKeyboard/Pinyin/SentenceLanguageModel.swift
import Foundation

/// 整句语言模型(2026-07,九宫格智能整句核心):词级 bigram(主)+ 字级 bigram OOV veto(兜底)。
///
/// 修"打对拼音却出弱智句"(因为你搬过来→因为你包裹来):原整句 DP 只按词频打分,无上下文语言模型。
/// 给整句 DP 的**跨词边界**打分:词 bigram 命中→中心化 delta(奖励好搭配),未命中→字级 OOV veto(仅罚)。
/// 参数与 tools/keyboard-verify/engine_lm.py 逐一对齐(全量回归电池 13/13 验证)。
///
/// 存储照抄 CustomDictionary/ByteTable:Data(.mappedIfSafe) 内存映射 + 行偏移索引 + 字节二分。
/// 数据文件行按 key 排序(生成期 sorted),可直接二分。命中行才转 String 解析权重(内存友好)。
final class SentenceLanguageModel {

    private struct Table {
        var bytes: Data = Data()
        var lineStart: [UInt32] = []
        var keyLen: [UInt16] = []
        var lineEnd: [UInt32] = []
        var ready = false
    }

    private var wordT = Table()
    private var charT = Table()
    private(set) var isReady = false

    func load() {
        charT = parse(resource: "char_bigram")
        wordT = parse(resource: "word_bigram")
        isReady = charT.ready || wordT.ready
    }

    private func parse(resource: String) -> Table {
        var t = Table()
        guard let url = Bundle(for: SentenceLanguageModel.self).url(forResource: resource, withExtension: "txt"),
              let data = try? Data(contentsOf: url, options: .mappedIfSafe) else { return t }
        var starts: [UInt32] = []; var klens: [UInt16] = []; var ends: [UInt32] = []
        let NL: UInt8 = 0x0A, TAB: UInt8 = 0x09
        data.withUnsafeBytes { (buf: UnsafeRawBufferPointer) in
            let n = buf.count
            var i = 0
            while i < n {
                let start = i
                var tab = -1
                var j = i
                while j < n && buf[j] != NL {
                    if buf[j] == TAB && tab < 0 { tab = j }
                    j += 1
                }
                if tab > start {
                    starts.append(UInt32(start)); klens.append(UInt16(tab - start)); ends.append(UInt32(j))
                }
                i = j + 1
            }
        }
        t.bytes = data; t.lineStart = starts; t.keyLen = klens; t.lineEnd = ends; t.ready = !starts.isEmpty
        return t
    }

    // MARK: 字节二分定位 key 行

    private func compareKey(_ t: Table, _ index: Int, _ q: [UInt8]) -> Int {
        let s = Int(t.lineStart[index]); let len = Int(t.keyLen[index])
        let m = min(len, q.count)
        var k = 0
        while k < m {
            let a = t.bytes[s + k], b = q[k]
            if a != b { return a < b ? -1 : 1 }
            k += 1
        }
        if len == q.count { return 0 }
        return len < q.count ? -1 : 1
    }

    /// 返回 key 精确命中的行号;无则 -1。
    private func findLine(_ t: Table, _ key: String) -> Int {
        if !t.ready { return -1 }
        let q = Array(key.utf8)
        var lo = 0, hi = t.lineStart.count
        while lo < hi {
            let mid = (lo + hi) >> 1
            if compareKey(t, mid, q) < 0 { lo = mid + 1 } else { hi = mid }
        }
        if lo < t.lineStart.count && compareKey(t, lo, q) == 0 { return lo }
        return -1
    }

    /// 行内查 key2 的权重(行体 `k2 w2 k3 w3 ...`)。命中返回权重,否则 nil。
    private func weightIn(_ t: Table, _ line: Int, _ target: String) -> Int? {
        let s = Int(t.lineStart[line]) + Int(t.keyLen[line]) + 1
        let e = Int(t.lineEnd[line])
        guard s < e else { return nil }
        let parts = String(decoding: t.bytes[s ..< e], as: UTF8.self).split(separator: " ")
        var i = 0
        while i + 1 < parts.count {
            if parts[i] == target { return Int(parts[i + 1]) }
            i += 2
        }
        return nil
    }

    // MARK: 打分(与 engine_lm.py 逐一对齐)

    /// 跨词边界打分:词 bigram 命中→中心化 delta;未命中→字级 OOV veto。整句 DP 每次词转移调用一次。
    func boundaryScore(_ pw: String, _ w: String) -> Int {
        if !isReady || pw.isEmpty || w.isEmpty { return 0 }
        if let wd = wordDelta(pw, w) { return wd }
        return charVeto(String(pw.last!), String(w.first!))
    }

    private func wordDelta(_ pw: String, _ w: String) -> Int? {
        let line = findLine(wordT, pw)
        if line < 0 { return nil }
        guard let b = weightIn(wordT, line, w) else { return nil }
        let x = Int(Double(b - Self.WREF) * Self.WBETA)
        if x > Self.WHI { return Self.WHI }
        if x < -Self.WLO { return -Self.WLO }
        return x
    }

    private func charVeto(_ prevLast: String, _ curFirst: String) -> Int {
        let line = findLine(charT, prevLast)
        if line < 0 { return -Self.LM_LO }
        guard let wv = weightIn(charT, line, curFirst) else { return -Self.LM_LO }
        let x = Int(Double(wv - Self.LM_REF) * Self.LM_BETA)
        if x >= 0 { return 0 }
        return x > -Self.LM_LO ? x : -Self.LM_LO
    }

    // 与 tools/keyboard-verify/engine_lm.py 逐一对齐(勿单独改,须同步 Python 参照并重跑回归)
    private static let WREF = -700
    private static let WBETA = 0.13
    private static let WHI = 55
    private static let WLO = 40
    private static let LM_REF = -520
    private static let LM_BETA = 0.16
    private static let LM_LO = 70
}

File: lobster-input-android/app/src/main/java/com/lobster/input/keyboard/pinyin/SentenceLanguageModel.kt
package com.lobster.input.keyboard.pinyin

import android.content.Context

/**
 * 整句语言模型(2026-07,九宫格智能整句核心):词级 bigram(主)+ 字级 bigram OOV veto(兜底)。
 *
 * 修复"打对拼音却出弱智句"(因为你搬过来→因为你包裹来):原整句 DP 只按词频打分,无上下文语言模型,
 * "包裹"高频词吃掉同码"搬过来"。本模型给整句 DP 的**跨词边界**打分:
 *  - 词级 bigram(主信号,精确):见过的词转移(堪称→完美 / 搬→过来 / 人民→是)给中心化 delta 奖励;
 *    字级恰好把这些判罕见(称→完 比 称→赞 罕见),故必须词级。数据 word_bigram.txt。
 *  - 字级 bigram OOV veto(兜底,仅罚):词 bigram 未覆盖的词对,用字级边界否决"从未出现的字搭配"
 *    (裹→来 类);只罚不奖,不引语料偏置。数据 char_bigram.txt。
 *
 * 只作用于**跨词边界**(词内已被词频编码);词频先验主导,LM 只在同码歧义时裁决。
 * 参数与 tools/keyboard-verify/engine_lm.py 逐一对齐(全量回归电池 13/13 验证)。
 *
 * 数据格式(与主词库同,UTF-8 文本):每行 `key\tk2 w2 k3 w3 ...`,key/k 为字或词,w 为量化整数。
 * 内存:HashMap 加载(char ~5.8K key,word ~35K key;权重存 Short 省内存),后台异步挂载不阻塞冷启动。
 */
class SentenceLanguageModel {

    // key1 -> (key2 -> 量化权重)。word 用词字符串,char 用单字字符串(统一 String,简化实现)。
    private var wordBi: HashMap<String, HashMap<String, Short>> = HashMap()
    private var charBi: HashMap<String, HashMap<String, Short>> = HashMap()
    @Volatile var isReady = false
        private set

    fun load(context: Context, charAsset: String = "char_bigram.txt", wordAsset: String = "word_bigram.txt") {
        charBi = parse(context, charAsset, 6_000)
        wordBi = parse(context, wordAsset, 40_000)
        isReady = true
    }

    private fun parse(context: Context, asset: String, cap: Int): HashMap<String, HashMap<String, Short>> {
        val map = HashMap<String, HashMap<String, Short>>(cap)
        context.assets.open(asset).bufferedReader(Charsets.UTF_8).useLines { lines ->
            for (line in lines) {
                val tab = line.indexOf('\t')
                if (tab <= 0) continue
                val k1 = line.substring(0, tab)
                val body = line.substring(tab + 1)
                val toks = body.split(' ')
                val inner = HashMap<String, Short>(toks.size / 2 + 1)
                var i = 0
                while (i + 1 < toks.size) {
                    val k2 = toks[i]
                    val w = toks[i + 1].toIntOrNull()
                    if (k2.isNotEmpty() && w != null) inner[k2] = w.coerceIn(-32768, 32767).toShort()
                    i += 2
                }
                if (inner.isNotEmpty()) map[k1] = inner
            }
        }
        return map
    }

    /**
     * 跨词边界打分:前词 pw → 当前词 w。词 bigram 命中用中心化 delta(可正可负),
     * 未命中回退字级 OOV veto(仅罚)。整句 DP 每次词转移调用一次。
     */
    fun boundaryScore(pw: String, w: String): Int {
        if (!isReady || pw.isEmpty() || w.isEmpty()) return 0
        val wd = wordDelta(pw, w)
        if (wd != null) return wd
        return charVeto(pw[pw.length - 1], w[0])
    }

    /** 词转移中心化 delta;未覆盖返回 null(交给字 veto)。 */
    private fun wordDelta(pw: String, w: String): Int? {
        val d = wordBi[pw] ?: return null
        val b = d[w] ?: return null
        val x = ((b - WREF) * WBETA).toInt()
        return when {
            x > WHI -> WHI
            x < -WLO -> -WLO
            else -> x
        }
    }

    /** 字级边界:只罚不奖,按罕见度分级(未见=满罚,罕见=轻罚,常见=0)。 */
    private fun charVeto(prevLast: Char, curFirst: Char): Int {
        val d = charBi[prevLast.toString()] ?: return -LM_LO
        val wv = d[curFirst.toString()] ?: return -LM_LO
        val x = ((wv - LM_REF) * LM_BETA).toInt()
        if (x >= 0) return 0
        return if (x > -LM_LO) x else -LM_LO
    }

    companion object {
        // 与 tools/keyboard-verify/engine_lm.py 逐一对齐(勿单独改动,须同步 Python 参照并重跑回归)
        private const val WREF = -700          // 词条件对数概率参考(约中等转移)
        private const val WBETA = 0.13         // 词 delta 斜率
        private const val WHI = 55             // 词 delta 正激励上限(突出好搭配)
        private const val WLO = 40             // 词 delta 负罚上限
        private const val LM_REF = -520        // 字 veto:高于此不罚
        private const val LM_BETA = 0.16       // 字 veto 斜率
        private const val LM_LO = 70           // 字 veto 单边界最大罚
    }
}
